sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Strategy.scala

```diff
@@ -52,21 +52,36 @@ class DataSourceV2Strategy(session: SparkSession) extends Strategy with Predicat
     }
   }
 
+  private def cache(relation: DataSourceV2Relation, cacheInfo: CacheInfo): Unit = {
+    session.sharedState.cacheManager.cacheQuery(
+      session,
+      relation,
+      cacheInfo.tableName,
+      cacheInfo.storageLevel)
+  }
+
   private def refreshCache(r: DataSourceV2Relation)(): Unit = {
     session.sharedState.cacheManager.recacheByPlan(session, r)
   }
 
-  private def invalidateCache(r: ResolvedTable, recacheTable: Boolean = false)(): Unit = {
+  // Invalidates the cache associated with the given table. If there exists a cache with the given
+  // table, the cache's info (table name and storage level) is returned.
+  private def invalidateCache(
+      r: ResolvedTable,
+      recacheTable: Boolean = false)(): Option[CacheInfo] = {
     val v2Relation = DataSourceV2Relation.create(r.table, Some(r.catalog), Some(r.identifier))
     val cache = session.sharedState.cacheManager.lookupCachedData(v2Relation)
     session.sharedState.cacheManager.uncacheQuery(session, v2Relation, cascade = true)
-    if (recacheTable && cache.isDefined) {
-      // save the cache name and cache level for recreation
+    if (cache.isDefined) {
       val cacheName = cache.get.cachedRepresentation.cacheBuilder.tableName
       val cacheLevel = cache.get.cachedRepresentation.cacheBuilder.storageLevel
```

Comment:
`cacheName` and `cacheLevel` is needed only in `if (recacheTable) { ... }` block.

Reply:
I think they are also needed for the returning `CacheInfo`.

Reply:
Oops, yes, You're right.